File: src/main/java/at/searles/regexp/Regexp.java
package at.searles.regexp;

/**
 * Interface for regular expressions
 */
public interface Regexp {

    static Regexp eof() {
        return CharSet.chars(-1);
    }

    // 'abc'
    static Regexp text(String s) {
        return new Text(s);
    }

    <A> A accept(Visitor<A> visitor);

    // A B
    default Regexp then(Regexp that) {
        return new Concat(this, that);
    }

    // A | B
    default Regexp or(Regexp that) {
        return new Union(this, that);
    }

    // A?
    default Regexp opt() {
        return new Closure(this, true, false);
    }

    // A*
    default Regexp rep() {
        return new Closure(this, true, true);
    }

    // A+
    default Regexp plus() {
        return new Closure(this, false, true);
    }

    // A{min,max}
    default Regexp range(int min, int max) {
        return new RepRange(this, min, max);
    }

    // A{min,}
    default Regexp min(int min) {
        return new RepMin(this, min);
    }

    // A{count}
    default Regexp count(int count) {
        return new RepCount(this, count);
    }

    // A^
    default Regexp nonGreedy() {
        return new NonGreedy(this);
    }
}

File: src/main/java/at/searles/regexp/Regexp.kt
package at.searles.regexp

/**
 * Interface for regular expressions
 */
interface Regexp {
    fun <A> accept(visitor: Visitor<A>): A

    // A B
    operator fun plus(that: Regexp): Regexp {
        return Concat(this, that)
    }

    // A | B
    infix fun or(that: Regexp): Regexp {
        return Union(this, that)
    }

    // A?
    fun opt(): Regexp {
        return Opt(this)
    }

    // A*
    fun rep(): Regexp {
        return KleeneStar(this)
    }

    // A+
    fun rep1(): Regexp {
        return KleenePlus(this)
    }

    // A{min,max}
    fun range(min: Int, max: Int): Regexp {
        return Range(this, min, max)
    }

    // A{min,}
    fun min(min: Int): Regexp {
        return AtLeast(this, min)
    }

    // A{count}
    fun count(count: Int): Regexp {
        return Count(this, count)
    }

    // A^
    fun nonGreedy(): Regexp {
        return FirstMatch(this)
    }

    infix fun and(other: Regexp): Regexp {
        return Intersect(this, other)
    }

    operator fun minus(other: Regexp): Regexp {
        return Subtract(this, other)
    }
}